packages/fxa-auth-server/lib/payments/apple-app-store/subscription-purchase.ts

```diff
@@ -0,0 +1,160 @@
+  private expirationIntent?: number;
+  private expiresDate?: number;
+  private isInBillingRetry?: boolean;
+  originalTransactionId: string; // unique identifier for the subscription; analogous to a Stripe subscription id
+  private productId: string; // unique identifier for the plan; analogous to the Stripe plan id
+  private status: SubscriptionStatus;
+  private type: TransactionType;
+
+  // Library-managed properties
+  userId?: string; // hex string for FxA user id
+  private verifiedAt: number; // timestamp of last purchase verification by App Store Server API
+
+  // Convert raw API response from App Store Server API to a SubscriptionPurchase object
+  static fromApiResponse(
+    apiResponse: StatusResponse,
+    subscriptionStatus: LastTransactionsItem['status'],
+    transactionInfo: JWSTransactionDecodedPayload,
+    renewalInfo: JWSRenewalInfoDecodedPayload,
+    originalTransactionId: string,
+    verifiedAt: number
+  ): SubscriptionPurchase {
+    const purchase = new SubscriptionPurchase();
+    purchase.autoRenewStatus = renewalInfo.autoRenewStatus;
+    purchase.bundleId = apiResponse.bundleId;
+    purchase.originalTransactionId = originalTransactionId;
+    purchase.productId = transactionInfo.productId;
+    purchase.status = subscriptionStatus;
+    purchase.type = transactionInfo.type;
+    purchase.verifiedAt = verifiedAt;
+
+    if (transactionInfo.expiresDate) {
+      purchase.expiresDate = transactionInfo.expiresDate;
+    }
+    if (renewalInfo.expirationIntent) {
+      purchase.expirationIntent = renewalInfo.expirationIntent;
+    }
+    if (renewalInfo.isInBillingRetryPeriod) {
+      // We don't check this.status === SubscriptionStatus.InBillingRetry, since
+      // it's not mutually exclusive with other subscription states (i.e.
+      // SubscriptionStatus.InBillingGracePeriod).
+      purchase.isInBillingRetry = renewalInfo.isInBillingRetryPeriod;
+    }
+
+    return purchase;
+  }
+
+  /**
+   * Converts purchase data from a Firestore document to a SubscriptionPurchase.
+   * In particular, there are some shopkeeping properties that are only used in
+   * Firestore; see FIRESTORE_OBJECT_INTERNAL_KEYS.
+   */
+  static fromFirestoreObject(firestoreObject: any) {
+    const purchase = new SubscriptionPurchase();
+    purchase.mergeWithFirestorePurchaseRecord(firestoreObject);
+    return purchase;
+  }
+
+  toFirestoreObject(): any {
+    return purchaseToFirestoreObject(this);
+  }
+
+  mergeWithFirestorePurchaseRecord(firestoreObject: any) {
+    mergePurchaseWithFirestorePurchaseRecord(this, firestoreObject);
+  }
+
+  /**
+   * Checks if a subscription is terminal -- i.e. immutable. In some situations,
+   * a new subscription is created rather than updating the original subscription.
+   */
+  isRegisterable(): boolean {
+    // Check if the subscription has lapsed, or we are no longer in the billing
+    // retry period
+    return this.isEntitlementActive() || this.isInBillingRetryPeriod();
+  }
```

Comment:
I'll take another pass through the slides on the Apple videos, otherwise we may have to check later with a test subscription that exhausts billing retry.

Reply:
A new subscription purchase would be needed if the original transaction id was different. That's the primary key of these, so if a purchase shares that, than its linked to the original so it updates the existing one, etc.